### ISSUE TITLE
Make sure engine restores original resolution if crashed in fullscreen

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#930]
According to previous @rofl0r 's comment in #23:

> i still get the occasional segfault from the engine though, which has the same effect. thus i don't run ags in fullscreen anymore. that particular issue could potentially be fixed though by using a signal handler for SIGSEGV that calls the "restore_original_resolution" code.

Need to make sure that engine catches exceptions and sigsev signals, and does a minimal cleanup before quitting.

**Comment by rofl0r** [comment]
if we add a sigsegv handler, than i'd vote for installing it only when screen resolution was changed, otherwise it will interfere with debugging in window mode (you can't really debug in FS mode anyway).

**Comment by rofl0r** [comment]
oh btw using signal handlers is generally problematic: only certain things may be done legally, one of the things that may not be done is allocating/freeing memory, so that's something to keep in mind.

**Comment by sonneveld** [comment]
> only certain things may be done legally, one of the things that may not be done is allocating/freeing memory, so that's something to keep in mind.

That and other things like printf I believe. It might be worth writing a special signal-only restore resolution function, but we'd have to check what allegro does.

**Comment by rofl0r** [comment]
> That and other things like printf I believe

yes, that's practically the same issue: printf uses stdout which uses buffered I/O, which may use allocations behind the curtains.

**Comment by sonneveld** [comment]
> yes, that's practically the same issue: printf uses stdout which uses buffered I/O, which may use allocations behind the curtains.

That's true, but non-reentrant functions may simply have a statically allocated structures. I think we're all in agreement that we should be careful not to use non-reentrant functions in a signal handler :)

Getting back to the resolution issue, does Linux support fullscreen modes like borderless fullscreen windows? That would allow AGS to effectively be full screen without changing the resolution.

**Comment by ivan-mogilko** [comment]
> Getting back to the resolution issue, does Linux support fullscreen modes like borderless fullscreen windows?

I think this is what it does for OpenGL.

**Comment by rofl0r** [comment]
> Getting back to the resolution issue, does Linux support fullscreen modes like borderless fullscreen windows?

that's possible but performance with 2D renderer is way worse with upscaled hi-res than real fullscreen mode.

> I think we're all in agreement that we should be careful not to use non-reentrant functions in a signal handler :)

let's just look at [the standard](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_04) and what it says in detail:

> the behavior is undefined if the signal handler refers to any object other than errno with static storage duration other than by assigning a value to an object declared as volatile sig_atomic_t, or if the signal handler calls any function defined in this standard other than one of the functions listed in the following table.

> The following table defines a set of functions that shall be async-signal-safe. Therefore, applications can call them, without restriction, from signal-catching functions. Note that, although there is no restriction on the calls themselves, for certain functions there are restrictions on subsequent behavior after the function is called from a signal-catching function (see longjmp). [list follows]

so in short, we may only call async-signal-safe functions, and access only errno and single atomic ints.

this makes it pretty hard to write a useful signal handler in this case. the only legal way i see here is to have a separate thread that periodically checks such a volatile sigatomic_t, set by another thread which runs the engine. the engine thread would set the flag if it crashes, or wants to restore the screen during exit.
i'm not sure whether that's worth the effort.

**Comment by morganwillcock** [comment]
I think in the majority of cases, a resolution change on a modern monitor is going to give a fuzzy image as the display panels are only designed to work with their native resolution in mind. On Windows 10 you also get hassled by pop-up notifications that you have incorrectly configured the display. To be honest I was going to suggest removing the resolution switching entirely and only provide scaling, in the long term.

> i'm not sure whether that's worth the effort.

I mention this here in-case it contributes to the decision.

**Comment by sonneveld** [comment]
@rofl0r It might be worth debugging the default signal handler that Allegro installs. It calls allegro_exit which shuts down the graphics module via function ptr. 

installing signals: https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/src/linux/lsystem.c#L202  
signal handler: https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/src/linux/lsystem.c#L129
shutdown_gfx, one of the functions added via _add_exit_func: https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/src/graphics.c#L456

`shutdown_gfx` calls `set_gfx_mode(GFX_TEXT...` which _should_ reset the desktop resolution.

**Comment by sonneveld** [comment]
> that's possible but performance with 2D renderer is way worse with upscaled hi-res than real fullscreen mode.

> To be honest I was going to suggest removing the resolution switching entirely and only provide scaling, in the long term.

I anticipate revisiting this when we move to a solely hardware accelerated renderer. (even if just textured quads in opengl)

**Comment by rofl0r** [comment]
> I think in the majority of cases, a resolution change on a modern monitor is going to give a fuzzy image as the display panels are only designed to work with their native resolution in mind

yeah, in some cases the image might not be super-crisp. but it's still a lot better to run a not-super-crisp game in fullscreen mode at 40 fps vs 10 fps at upscaled.

> To be honest I was going to suggest removing the resolution switching entirely and only provide scaling, in the long term.

this seems really short-sighted and unfair to users that don't have top of the line hardware. i don't think this minor issue justifies ripping out fullscreen support.

> I anticipate revisiting this when we move to a solely hardware accelerated renderer. (even if just textured quads in opengl)

the software renderer is there for a good reason: not everybody has top 3D hardware, especially linux users that often have no proper upstream driver support and require hackish reverse-engineered drivers like nouveau or don't have any GL drivers at all.

**Comment by rofl0r** [comment]
> @rofl0r It might be worth debugging the default signal handler that Allegro installs. It calls allegro_exit which shuts down the graphics module via function ptr.

yeah, that's an option. though i think this should be explored the next time it happens. since looking at this code, there's already a SIGSEGV handler so it might have been fixed already on the allegro side.

**Comment by ivan-mogilko** [comment]
> the software renderer is there for a good reason

To be fair, software renderer is there simply because it was the only renderer created originally.
As time goes it becomes harder to keep supporting it, because for each graphic effect we add there has to be a software equivalent, and it often requires special approach to optimizing drawing operations.

**Comment by rofl0r** [comment]
> Frankly, as time goes it becomes harder to keep supporting it, because for each graphic effect we add there has to be a software equivalent

are new GFX still added regularly or is this something you envision for ags4, where you may want to use shaders and stuff like that in a GL-only renderer ?

**Comment by ericoporto** [comment]
I had an idea to make exclusive fullscreen always use the display resolution that is set, in that way any crash has no impact in resolution.
